File: adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/importer/Importer.java
package org.jvalue.ods.adapterservice.adapter.importer;

import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonProperty;

public abstract class Importer {

    public abstract String getType();
    public abstract String getDescription();

    @JsonProperty("parameters")
    public abstract List<ImporterParameterDescription> getAvailableParameters();

    protected List<ImporterParameterDescription> getRequiredParameters() {
      return getAvailableParameters();
    }

    public final String fetch(Map<String, Object> parameters) {
      validateParameters(parameters);
      return doFetch(parameters);
    }
    protected abstract String doFetch(Map<String, Object> parameters);

    protected void validateParameters(Map<String, Object> inputParameters) {
      boolean illegalArguments = false;
      String illegalArgumentsMessage = "";

      for (ImporterParameterDescription requiredParameter : getRequiredParameters()) {
        if (inputParameters.get(requiredParameter.getName()) == null) {
          illegalArguments = true;
          illegalArgumentsMessage = illegalArgumentsMessage + getType() + " importer requires parameter "
            + requiredParameter.getName() + "/n";

        } else if (inputParameters.get(requiredParameter.getName()).getClass() != requiredParameter.getType()) {
          illegalArguments = true;
          illegalArgumentsMessage = illegalArgumentsMessage + getType() + " importer requires parameter "
            + requiredParameter.getName() + " to be type " + requiredParameter.getType().toString() + "/n";
        }
      }
      if (illegalArguments) {
        throw new IllegalArgumentException(illegalArgumentsMessage);
      }
    }

}

File: adapter/src/adapter/importer/Importer.ts
import { ImporterParameterError } from '../model/exceptions/ImporterParameterError';

export interface ImporterParameterDescription {
  name: string;
  description: string;
  required: boolean;
  type: unknown;
}

export abstract class Importer {
  constructor(public type: string, public description: string) {}

  /**
   * @returns a list of all available parameters for this importer
   */
  abstract getAvailableParameters(): ImporterParameterDescription[];

  /**
   * Actual logic to fetch the data.
   * @returns the imported data as string
   */
  abstract doFetch(parameters: Record<string, unknown>): Promise<string>;

  getRequiredParameters(): ImporterParameterDescription[] {
    return this.getAvailableParameters().filter(
      (item: ImporterParameterDescription) => item.required,
    );
  }

  /**
   * Validates the given parameters and imports data as string
   */
  async fetch(parameters: Record<string, unknown>): Promise<string> {
    this.validateParameters(parameters);
    const x = await this.doFetch(parameters);
    return x;
  }

  /**
   * Validates the input parameters (Generic function, used in the derived classes)
   *
   * @param inputParameters the parameters to be validated. Checks if there are all required parameters provided and the type is correct.
   * Also checks if there are unnecessary arguments provided.
   *
   * @returns void
   * @throws InterpreterParameterError, if an error is found
   */
  validateParameters(inputParameters: Record<string, unknown>): void {
    let illegalArguments = false;
    let illegalArgumentsMessage = '';

    const possibleParameters: ImporterParameterDescription[] =
      this.getAvailableParameters();

    const unnecessaryArguments = [];
    const names = possibleParameters.map((a) => a.name);
    const keys = Object.keys(inputParameters);

    for (const entry of keys) {
      if (!names.includes(entry)) {
        unnecessaryArguments.push(entry);
      }
    }

    if (unnecessaryArguments.length > 0) {
      illegalArguments = true;
      for (const argument of unnecessaryArguments) {
        illegalArgumentsMessage += argument + ' is not needed by importer \n';
      }
    }

    const requiredParameters = this.getRequiredParameters();
    for (const requiredParameter of requiredParameters) {
      const param = inputParameters[
        requiredParameter.name
      ] as ImporterParameterDescription;

      if (inputParameters[requiredParameter.name] === undefined) {
        illegalArguments = true;
        illegalArgumentsMessage += this.type;
        illegalArgumentsMessage += 'importer requires parameter ';
        illegalArgumentsMessage += requiredParameter.name;
        illegalArgumentsMessage += '\n';
        break;
      }
      const checkType = param.constructor.name;
      if (checkType.toLowerCase() !== requiredParameter.type) {
        illegalArguments = true;
        illegalArgumentsMessage += this.type;
        illegalArgumentsMessage += ' importer requires parameter ';
        illegalArgumentsMessage += requiredParameter.name;
        illegalArgumentsMessage += ' to be type ';
        illegalArgumentsMessage += requiredParameter.type;
        illegalArgumentsMessage += '\n';
        break;
      }
    }
    if (illegalArguments) {
      throw new ImporterParameterError(illegalArgumentsMessage);
    }
  }
}
